sys/net/application_layer/gcoap/dns.c

```diff
@@ -0,0 +1,683 @@
+    }
+    assert(creds != NULL);
+#if IS_USED(MODULE_GCOAP_DTLS)
+    for (unsigned i = 0; i < ARRAY_SIZE(_creds); i++) {
+        if (!c &&
+            (((_creds[i].type == creds->type) && (_creds[i].tag == creds->tag)) ||
+             (_creds[i].type == CREDMAN_TYPE_EMPTY))) {
+            c = &_creds[i];
+            break;
+        }
+    }
+#endif
+    if (c == NULL) {
+        DEBUG("gcoap_dns: no space to store credential tag\n");
+        return -ENOMEM;
+    }
+    int res = credman_add(creds);
+    if ((res < 0) && (res != CREDMAN_EXIST)) {
+        /* ignore duplicate credentials */
+        DEBUG("gcoap_dns: cannot add credential to system: %d\n", res);
+        return -EBADF;
+    }
+    if (res == CREDMAN_OK) {
+#if IS_USED(MODULE_GCOAP_DTLS)
+        /* functions used in here are only available with module gcoap_dtls, so guard this
+         * section */
+        sock_dtls_t *gcoap_sock_dtls = gcoap_get_sock_dtls();
+
+        res = sock_dtls_add_credential(gcoap_sock_dtls, creds->tag);
+        if (res < 0) {
+            DEBUG("gcoap_dns: cannot add credential to GCoAPs DTLS sock\n");
+            return -EBADF;
+        }
+#endif
+        c->tag = creds->tag;
+        c->type = creds->type;
+    }
+    return 0;
+}
+
+void gcoap_dns_cred_remove(credman_tag_t tag, credman_type_t type)
+{
+#if IS_USED(MODULE_GCOAP_DTLS)
+    sock_dtls_t *sock = gcoap_get_sock_dtls();
+    for (unsigned i = 0; i < ARRAY_SIZE(_creds); i++) {
+        if ((_creds[i].tag == tag) && (_creds[i].type == type)) {
+            _remove_cred(sock, &_creds[i]);
+            break;
+        }
+    }
+#else
+    (void)tag;
+    (void)type;
+#endif
+}
+
+void gcoap_dns_server_proxy_reset(void)
+{
+    if (IS_USED(MODULE_GCOAP_DNS_PROXIED)) {
+        mutex_lock(&_client_mutex);
+        _proxy[0] = '\0';
+        mutex_unlock(&_client_mutex);
+    }
+}
+
+int gcoap_dns_server_proxy_set(const char *proxy)
+{
+    int res;
+
+    if (!IS_USED(MODULE_GCOAP_DNS_PROXIED)) {
+        return -ENOTSUP;
+    }
+    res = strlen(proxy);
+
+    if ((unsigned)(res + 1) >= CONFIG_GCOAP_DNS_SERVER_URI_LEN) {
+        return -EINVAL;
+    }
+    mutex_lock(&_client_mutex);
+    strcpy(_proxy, proxy);
+    mutex_unlock(&_client_mutex);
+    return res;
+}
+
+const char *gcoap_dns_server_proxy_get(void)
+{
+    if (_is_proxied()) {
+        return _proxy;
+    }
+    return NULL;
+}
+
+static inline bool _dns_server_uri_isset(void)
+{
+    return _uri[0] != '\0';
+}
+
+#if IS_USED(MODULE_GCOAP_DTLS)
+static void _remove_cred(sock_dtls_t *sock, _cred_t *cred)
+{
+    sock_dtls_remove_credential(sock, cred->tag);
+    credman_delete(cred->tag, cred->type);
+    cred->type = CREDMAN_TYPE_EMPTY;
+    cred->tag = CREDMAN_TAG_EMPTY;
+}
+#endif
+
+static inline bool _is_proxied(void)
+{
+    return IS_USED(MODULE_GCOAP_DNS_PROXIED) && _proxy[0] != '\0';
+}
+
+static int _add_init_block2_opt(coap_pkt_t *pdu)
+{
+    if (CONFIG_GCOAP_DNS_PDU_BUF_SIZE < CONFIG_DNS_MSG_LEN) {
```

Comment:
```suggestion
    if (CONFIG_GCOAP_DNS_PDU_BUF_SIZE < CONFIG_DNS_MSG_LEN) {
        /* If our largest DNS message fits in the pbuf, there is no point in sending Block2
         * in the initial message -- a huge response might still overwhelm our PDU buffer,
         * but if that happens we could not have processed it as a DNS message if it came
         * in in fragments either. */
```

(Just because it took me quite a while to figure out that this is indeed correct)